Extend child-support

lhaagsma opened the pull request:
This PR adds a --list-children flag to target-query similar to https://github.com/fox-it/acquire/issues/239. 

Test output looks like this:

```
$ target-query --list-children <target>
Processing target: local (hostname=pve)
- [#0]: type=proxmox, path=/etc/pve/qemu-server/101.conf
- [#1]: type=proxmox, path=/etc/pve/qemu-server/102.conf
- [#2]: type=proxmox, path=/etc/pve/qemu-server/103.conf
- [#3]: type=proxmox, path=/etc/pve/qemu-server/100.conf
- [#4]: type=proxmox, path=/etc/pve/qemu-server/104.conf
```


Additionally, added index to the target-info plugin.

```
$ target-info <target>
<snip>
Children
- <Child #="0" type="proxmox" path="/etc/pve/qemu-server/101.conf">
- <Child #="1" type="proxmox" path="/etc/pve/qemu-server/102.conf">
- <Child #="2" type="proxmox" path="/etc/pve/qemu-server/103.conf">
- <Child #="3" type="proxmox" path="/etc/pve/qemu-server/100.conf">
- <Child #="4" type="proxmox" path="/etc/pve/qemu-server/104.conf">
<snip>

```

Aims to close: #1132 and #1134

Schamper reviewed (changes requested):
Can you add a unit test?

And @JSCU-CNI not sure if the # key name breaks anything?

lhaagsma commented:
The `#` or index (index is reserved in Splunk..) is mainly useful for cli purposes and seems pretty 'relative' to the target. Might not be too valuable to include in the records. 
Instead i'm looking into retrieving the child.name from config files which seems more valuable.

lhaagsma commented:
So i'm struggling a bit on how to do this in target-info. 
I've added (and i'm about to commit an initial version)  some code to target-info but the overall structure of target-info does not make it easy to pass arguments to get_target_info and/or print_target_info.
I have added code --list-children and --list-children-recursive to the other tools (query/shell) and will also be added to acquire.

lhaagsma commented:
Latest commit adds "--list-children" and "--list-children-recursive" to `target-query` and `target-shell`. 
A start was made for `target-info` but looking for guidance how to properly pass an optional parameter for recursive to the `get_child_info` function. @JSCU-CNI / @Schamper ?

`target-shell` was expanded to allow for --child to select a (sub)child from a system. example: `target-shell --child 4.1`

Example output for --list-children-recursive:
```
Processing hostname=pve, path=local (-recursive=True)
 [#0]: type=proxmox, name=wireguard, path=/etc/pve/qemu-server/101.conf
 [#1]: type=proxmox, name=proxy, path=/etc/pve/qemu-server/102.conf
 [#2]: type=proxmox, name=vm1, path=/etc/pve/qemu-server/103.conf
 [#3]: type=proxmox, name=vm2, path=/etc/pve/qemu-server/100.conf
 [#4]: type=proxmox, name=docker-host, path=/etc/pve/qemu-server/104.conf
 -[#4.0]: type=docker, name=angry_sammet, path=/var/lib/docker/image/overlay2/layerdb/mounts/xxx
 -[#4.1]: type=docker, name=changedetectionio-playwright-chrome-1, path=/var/lib/docker/image/overlay2/layerdb/mounts/xxx
 -[#4.2]: type=docker, name=docker1, path=/var/lib/docker/image/overlay2/layerdb/mounts/xxx
 -[#4.3]: type=docker, name=docker2-grafana-1, path=/var/lib/docker/image/overlay2/layerdb/mounts/xxxx
 -[#4.4]: type=docker, name=docker3-mosquitto-1, path=/var/lib/docker/image/overlay2/layerdb/mounts/xxxx
 -[#4.5]: type=docker, name=docker4-1, path=/var/lib/docker/image/overlay2/layerdb/mounts/xxx
 -[#4.6]: type=docker, name=docker5-database-1, path=/var/lib/docker/image/overlay2/layerdb/mounts/xxx
 [#5]: type=proxmox, name=VMWarez, path=/etc/pve/qemu-server/106.conf
 [#6]: type=proxmox, name=Win11hyperrrrrV, path=/etc/pve/qemu-server/105.conf
 -[#6.0]: type=hyper-v, name=VirtualOffspring, path=C:\ProgramData\Microsoft\Windows\Hyper-V\Virtual Machines\<guid>.vmcx
 --[#6.0.0]: type=virtualbox, name=test_vm, path=/home/user/VirtualBox VMs/test_vm/test_vm.vbox
 -[#6.1]: type=wsl, name=kali-linux, path=C:\Users\user\AppData\Local\wsl\{guid}\ext4.vhdx
 -[#6.2]: type=wsl, name=Debian, path=C:\Users\user\AppData\Local\wsl\{guid}\ext4.vhdx
 -[#6.3]: type=wsl, name=name_name, path=C:\Users\user\AppData\Local\wsl\{guid}\ext4.vhdx
```


Additional the `ChildTargetRecord` was expanded to include the `name` field. This name field is now parsed for the following hypervisors:
- 	child/docker.py
- 	child/esxi.py   - not tested yet
- 	child/hyperv.py
- 	child/parallels.py  - not tested yet
- 	child/proxmox.py
- 	child/qemu.py    - not tested yet
- 	child/virtualbox.py
- 	child/virtuozzo.py    - not tested yet
- 	child/vmware_workstation.py    - not tested yet
- 	child/wsl.py

Not yet for: 
- child/colima.py

I currently do not have a working setup for all hypervisors so not al code has been tested to work on real (live) installations/systems. But all tox tests are currently passing. 

Todo:

- Add support for colima.py
- Commit acquire --list-children(-recursive).
- Extend unit tests to check for child names.
- Improve the error handling of the various _get_child_name functions within the child plugins. Any advice @Schamper ?

lhaagsma commented:
@Zawadidone any chance you could have a look/share how to retrieve the container name for Colima? Additionally, if you happen to have a test environment for virtuozzo and could validate that my changes work that would be great :)

lhaagsma commented:
Requesting review.

Schamper commented on dissect/target/plugins/apps/container/container.py:
```diff
@@ -22,7 +22,7 @@
     ("datetime", "finished"),
     ("string[]", "ports"),
-    ("string", "names"),
+    ("string", "name"),
```
@JSCU-CNI I changed this from `names` to `name`, seemed to be an oversight in naming since it's seemingly always a singular name, but let me know if I'm overlooking anything.

Schamper commented on dissect/target/plugins/child/wsl.py:
```diff
@@ -31,9 +31,15 @@ def find_wsl_installs(target: Target) -> Iterator[Path]:
                 base_path = target.resolve(distribution_key.value("BasePath").value)
                 # WSL needs diskname to be ext4.vhdx, but they can be renamed when WSL is not active
-                yield from base_path.glob("*.vhdx")
+                if not (disk_path := next(base_path.glob("*.vhdx"), None)):
```
Apparently not, @lhaagsma?

lhaagsma replied:
By default a WSL is downloaded/started it has one disk, so that should probably be sufficient for now.

Schamper commented on dissect/target/plugins/child/colima.py:
```diff
@@ -13,42 +13,56 @@
+            lima_id = f"colima-{profile}"
+
+        if (lima_path := path.joinpath("_lima", lima_id)).exists():
+            yield profile, config_file, lima_path
```
Just makes the function more complete and compatible for a future Colima config parsing plugin.
